lib/scheduler.ml

```diff
@@ -0,0 +1,198 @@
+(*
```

Comment:
This module should have an interface. It contains quite complicated code to implement an abstraction that clients should not be able to break.